proposals/NNNN-custom-isolation-checking-for-serialexecutor.md

```diff
@@ -0,0 +1,190 @@
+# Custom isolation checking for SerialExecutor
+
+* Proposal: [SE-NNNN](NNNN-advanced-custom-isolation-checking-for-serialexecutor.md)
+* Author: [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+* Review Manager: ???
+* Status:  **Work in Progress**
+* Implementation: [PR #71172](https://github.com/apple/swift/pull/71172)
+* Review: ???
+
+## Introduction
+
+Swift introduced custom actor executors in [SE-0392: Custom Actor Executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md), and ever since allowed further customization of isolation and execution semantics of actors.
+
+This proposal also introduced a family of assertion and assumption APIs which are able to dynamically check the isolation of a currently executing task. These APIs are:
+
+- Asserting isolation context:
+  - [`Actor/assertIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/assertisolated(_:file:line:))
+  - [`Actor/preconditionIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/preconditionisolated(_:file:line:))
+  - [`DistributedActor/assertIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/preconditionisolated(_:file:line:))
+  - [`DistributedActor/preconditionIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/preconditionisolated(_:file:line:))
+- Assuming isolation context, and obtaining an `isolated actor` reference of the target actor
+  - [`Actor/assumeIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/assumeisolated(_:file:line:))
+  - [`DistributedActor/assumeIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/assumeisolated(_:file:line:))
+
+## Motivation
+
+All the above mentioned APIs rely on an internal capability of the Swift concurrency runtime to obtain the "current serial executor", and compare it against the expected executor. Additional comparison modes such as "complex equality" are also supported, which help executors that e.g. share a single thread across multiple executor instances to still be able to correctly answer the "are we on the same executor?" question when different executor *instances* are being compared, however in reality they utilize the same threading resource.
+
+The proposal did not account for the situation in which the Swift concurrency runtime has no notion of "current executor" though, causing the following situation to -- perhaps surprisingly -- result in runtime crashes reporting an isolation violation, while in reality, no such violation takes place in the following piece of code:
```

Comment:
I think you can tell a somewhat more focused narrative here.  I'd suggest a structure something like this:
- The runtime already needs to dynamically track a task's current executor.
- `assumeIsolated` etc. are primarily built on top of that.
- That's good enough for uses from tasks, like when a task calls a synchronous non-isolated function.
- It's not good enough when code can be scheduled on the executor in other ways.
- The actor is still isolated in these other cases, but the runtime doesn't know about it because the scheduling didn't come through the runtime.
- This was a common and important enough situation that we added a special case for `MainActor`.
- But it also comes up with `DispatchQueue` and other serial executors, and these executors generally offer their own ways to assert that you're on the right executor.
- It makes sense to generalize that by allowing the executor to provide its own logic for safely asserting the precondition.

I don't think stuff like complex equality is really important to include.

Reply:
That sounds good, I'll rephrase the narrative -- it's good to compare this to our special casing of the main actor 👍